### ISSUE TITLE
Support the `interval` type when appropriate Rust types are developed

### ISSUE THREAD

**Issue by mikedilger** [sfackler/rust-postgres#60]
Is there a reason { name: date, oid: 1082 } is not included?  I realize a timestamp is more precise, but (at the very least) existing databases may use date.

I tried to add support for it at https://github.com/mikedilger/rust-postgres/commit/7f4637956f65c9086705c444ba93d30c4f593177 but I'm not sure that's right.  When I try using it, it triggers a RefCell<T> already borrowed error (#34 or https://github.com/rust-lang/rust/issues/13246)

**Comment by cbruun** [comment]
Is [rust-chrono](https://github.com/lifthrasiir/rust-chrono) a good choice?

**Comment by sfackler** [comment]
I've added support mapping `TIMESTAMP WITH TIME ZONE` to `DateTime<UTC>`, `TIMESTAMP` to `NaiveDateTime`, `DATE` to `NaiveDate` and `TIME` to `NaiveTime` under the `chrono` feature. There's still no equivalent for `INTERVAL` as far as I know.

**Comment by TheServerAsterisk** [comment]
@sfackler would there be interest if I wrote a "simple" mapping for the interval type? More specifically it would have the following interface and other additions if required: 

```rust
pub fn new(microseconds: i64, days: i32, months: i32) -> Interval;
pub fn microseconds(&self) -> i64;
pub fn days(&self) -> i32;
pub fn months(&self) -> i32;
pub fn to_sql_standard(&self) -> String; 
pub fn to_postgres(&self) -> String; 
pub fn to_postgres_verbose(&self) -> String; 
pub fn to_iso_8601(&self) -> String; 
```

and will also implement the `Clone`, `Copy`, `PartialEq`, `Eq`, `Debug`, `Add`, and `Sub` traits. 

Any input would be appreciated.

**Comment by voronaam** [comment]
This affects me as well. My current workaround is

```
conn.execute("DELETE FROM test WHERE ts < now() - $1::text::interval",
                    &[&format!("{}ms", retention)]);
```

The `::text` part is required for the library to map the type and it is later converted into an interval on the postgres side.

The @TheServerAsterisk suggestion makes sense to me. Though I'd suggest having a `new` variant which takes in `std::time::Duration`. But this is not important.

**Comment by piperRyan** [comment]
I have a dedicated interval type [here](https://crates.io/crates/pg_interval) if anyone is interested.

**Comment by jcaesar** [comment]
How foolish is it to wish for `std::time::Duration` to be converted to intervals?

**Comment by sfackler** [comment]
A Duration could be converted to a Postgres interval, but not vice versa: https://github.com/sfackler/rust-postgres/issues/60#issuecomment-57099402.

**Comment by sanpii** [comment]
> There's still no equivalent for INTERVAL as far as I know.

`chrono::Duration`?

**Comment by sfackler** [comment]
As noted in the comment directly above yours, those types aren't equivalent: https://github.com/sfackler/rust-postgres/issues/60#issuecomment-57099402

**Comment by piperRyan** [comment]
@sanpii I wrote a crate that is direct binding for interval type see my comment above. 

@sfackler if you would like the code I have so far added this crate please let me know.

**Comment by aloucks** [comment]
> We can easily implement ToSql for Duration, but any FromSql implementation would be very wrong.

Is there any reason why we can't have the `ToSql` trait implemented even if `FromSql` isn't possible? It would be really nice to be able to use `std::time::Duration` for things like:

```sql
SELECT foo_id FROM foo WHERE foo.expire_ts > CURRENT_TIMESTAMP + INTERVAL $1
```

**Comment by LukasKalbertodt** [comment]
I attempted to add `ToSql for Duration` but I didn't get far, mostly because I was unable to find good documentation. I'm leaving some notes and questions in this thread in order to hopefully help the next one attempting this.

---

Above and in the docs, it is mentioned that intervals consists of months, days and microseconds. Further, `interval` is 16 bytes large and months & days are ints. So I assume it's something like this?

```rust
// Maybe?
struct Interval {
    months: i32,
    days: i32,
    microseconds: i64,
}
```

But I'm not sure and I couldn't find any information on how those are written to the binary buffer. I briefly tried to dig through the postgres code, but gosh, I haven't read C code in a long time. And I just couldn't find anything relevant.

Finally, how would one convert `Duration` into `Interval`? @sfackler you said:

> We can easily implement `ToSql` for `Duration`

But how exactly? Especially since days and months vary in length (the whole point of this weird representation), one surely couldn't do any `seconds % 24 * 60 * 60` math, right? And in that case the `seconds: u64` from `Duration` cannot be represented by `microseconds: i64`, right?

**Comment by sanpii** [comment]
> So I assume it's something like this?

Yes, if you prefer I did the job for my crate: https://github.com/elephantry/elephantry/blob/3.0.0-beta.2/core/src/sql/date/interval.rs

**Comment by ctron** [comment]
I was running into the same limitation. Using `chrono::Duration`, is it possible to transform the duration into an ISO 8609 duration (like `PT1S` = `1 second`), this can be fed into a statement using the type varchar and converting it to an interval in the statement. Like `$1::interval`.

Not sure if that is a recipe to implement this out-of-the-box or even the other way around (from postgres to Duration). But it worked for me.

**Comment by auyer** [comment]
Sorry for pinging, but what is the best approach to solve this? 
I see that there are crates referenced in this issue that attempt to fix this issue. Is this enough ? 
If not, what is ? `chrono::Duration`  support ? Official Rust type support ?  If any of there, does issues about this exist ? 
Thanks!

**Comment by hongquan** [comment]
There is a crate, [chrono-intervals](https://crates.io/crates/chrono-intervals) that provides `Interval` type.

**Comment by levkk** [comment]
I think going the way SQLx did it with [`PgInterval`](https://docs.rs/sqlx/latest/sqlx/postgres/types/struct.PgInterval.html) isn't a bad idea, just so we can use this data type in Rust code.

**Comment by pnathan** [comment]
Is there a way that rust-postgres could have a custom type for `INTERVAL` so that we could read/write them and perform the transform on our (appdev) side?
